File: model/src/main/java/com/github/fburato/highwheelmodules/model/modules/ModuleGraph.java
package com.github.fburato.highwheelmodules.model.modules;

import java.util.Collection;
import java.util.Optional;

public interface ModuleGraph<T> {
  Optional<T> findDependency(HWModule vertex1, HWModule vertex2);

  void addDependency(T dependency);

  void addModule(HWModule vertex);

  Collection<HWModule> modules();

  Collection<HWModule> dependencies(HWModule vertex);
}

File: model/src/main/scala/com/github/fburato/highwheelmodules/model/modules/ModuleGraph.scala
package com.github.fburato.highwheelmodules.model.modules

trait ModuleGraph[T] {
  def findDependency(vertex1: HWModule, vertex2: HWModule): Option[T]

  def addDependency(dependency: T): Unit

  def addModule(vertex: HWModule): Unit

  def modules: Seq[HWModule]

  def dependencies(vertex: HWModule): Seq[HWModule]
}
